pkg/clients/compute/agentpool/agentpool.go

```diff
@@ -0,0 +1,152 @@
+	}
+	if len(c.Spec.AvailabilityZones) > 0 {
+		p.AvailabilityZones = &c.Spec.AvailabilityZones
+	}
+	if c.Spec.NodeCount != nil {
+		p.ManagedClusterAgentPoolProfileProperties.Count = c.Spec.NodeCount
+	}
+	if c.Spec.MinNodeCount != nil {
+		p.ManagedClusterAgentPoolProfileProperties.MinCount = c.Spec.MinNodeCount
+	}
+	if c.Spec.MaxNodeCount != nil {
+		p.ManagedClusterAgentPoolProfileProperties.MaxCount = c.Spec.MaxNodeCount
+	}
+	if c.Spec.VnetSubnetID != "" {
+		p.VnetSubnetID = to.StringPtr(c.Spec.VnetSubnetID)
+	}
+	if len(c.Spec.NodeTaints) > 0 {
+		p.NodeTaints = &c.Spec.NodeTaints
+	}
+	return p
+}
+
+// NewProfile returns an Azure ManagedClusterAgentPoolProfile object from a AKSCluster spec
+func NewProfile(c *v1alpha3.AKSCluster) containerservice.ManagedClusterAgentPoolProfile {
+	nodeCount := int32(v1alpha3.DefaultNodeCount)
+	if c.Spec.NodeCount != nil {
+		nodeCount = int32(*c.Spec.NodeCount)
+	}
+	p := containerservice.ManagedClusterAgentPoolProfile{
+		Name:   to.StringPtr(consts.AgentPoolProfileName),
+		Count:  &nodeCount,
+		VMSize: containerservice.VMSizeTypes(c.Spec.NodeVMSize),
+		Mode:   containerservice.System,
+		Type:   containerservice.VirtualMachineScaleSets,
+	}
+	if c.Spec.VnetSubnetID != "" {
+		p.VnetSubnetID = to.StringPtr(c.Spec.VnetSubnetID)
+	}
+	return p
+}
+
+// NeedUpdate determines if a AgentPool need to be updated
+func NeedUpdate(c *v1alpha3.AgentPool, az *containerservice.AgentPool) bool {
+	if c.Spec.NodeVMSize != string(az.VMSize) {
+		return true
+	}
+	if nodeCountNeedUpdate(c, az) {
+		return true
+	}
+	if c.Spec.VnetSubnetID != "" {
+		azureValue := azure.ToString(az.VnetSubnetID)
+		specValue := c.Spec.VnetSubnetID
+		if azureValue != specValue {
+			return true
+		}
+	}
+	if c.Spec.EnableAutoScaling != azure.ToBool(az.EnableAutoScaling) {
+		return true
+	}
+	if az.NodeTaints != nil {
+		azTaints := *az.NodeTaints
+		if !cmp.Equal(azTaints, c.Spec.NodeTaints) {
+			return true
+		}
+	}
+	if az.NodeTaints == nil && len(c.Spec.NodeTaints) > 0 {
+		return true
+	}
+	return false
+}
+
+func nodeCountNeedUpdate(c *v1alpha3.AgentPool, az *containerservice.AgentPool) bool {
+	if c.Spec.NodeCount != nil && c.Spec.MinNodeCount == nil && c.Spec.MaxNodeCount == nil {
+		azureCount := azure.ToInt(az.Count)
+		specCount := azure.ToInt(c.Spec.NodeCount)
+		if azureCount != specCount {
+			return true
+		}
+	}
+	if c.Spec.MinNodeCount != nil {
+		azureCount := azure.ToInt(az.MinCount)
+		specCount := azure.ToInt(c.Spec.MinNodeCount)
+		if azureCount != specCount {
+			return true
+		}
+	}
+	if c.Spec.MaxNodeCount != nil {
+		azureCount := azure.ToInt(az.MaxCount)
+		specCount := azure.ToInt(c.Spec.MaxNodeCount)
+		if azureCount != specCount {
+			return true
+		}
+	}
+	return false
+}
+
+// Update updates the status related to the external
+// Azure virtual network in the AgentPoolStatus
+func Update(c *v1alpha3.AgentPool, az *containerservice.AgentPool) {
```

Comment:
nit: Maybe renamed as `UpdateStatus` not to be confused with an infrastructure object update.